pytorch_lightning/metrics/metric.py

```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
         """
         super().__init__()
         self.name = name
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
+
+        # Register hooks
+        self.register_forward_pre_hook(self.input_convert)
+        self.register_forward_hook(self.output_convert)
+        self.register_forward_hook(self.ddp_sync)
+        self.register_forward_hook(self.compute)
 
     @abstractmethod
-    def forward(self, *args, **kwargs) -> torch.Tensor:
+    def forward(self, *args, **kwargs):
         """
         Implements the actual metric computation.
 
         Returns:
-            metric value
+            metric value or metric state
 
         """
         raise NotImplementedError
 
+    def compute(self, module: nn.Module, data: Any, output: Any):
+        """
+        Implement additionally metric computations to be done after the ddp sync
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+            output: output from forward method
```

Comment:
where do you get the output from? Shouldn't this be just computed in here?

Reply:
The output is whatever is returned by the forward method the user have implemented. Thus, if the user returns some inter median values (like number of samples and the squared error in my written example) these are reduced by the ddp_sync hook and when we reach this method we can to the final computations to get the correct value